### DISCUSSION TITLE
Support for Blazemeter  -report option

### DISCUSSION THREAD

**Discussion by VaishaliGulve** [abstracta/jmeter-java-dsl#185] [Q&A] [answered]
Hello,

I see that we have an option to run distributed tests directly from DSL. Do we also have an option to send the report to Blazemeter report? Currently , the Taurus wrapper provides that and we use that feature extensively. Can you please provide support for this feature?

**Comment by rabelenda** [comment]
Really interesting feature, thank you for asking and proposing it!

What do you think of something like this?:

```java
testPlan(
        threadGroup(2, 10,
            httpSampler("http://my.service")
        ),
        blazeMeterReportUploader()
    )
```

And that listener can be part of existing jmeter-java-dsl-blazemeter module.

Any particular fields you usually set with taurus blazemeter reports? 
I guess you usually set the BM token. Do you set report names or other fields?

**Reply by VaishaliGulve** [reply]
Thank you! Yes, BM token, test name, project name, report name these are the main fields. Also, currently it can be used in any of the below scenarios - 1. Distributed test ( where the test is internal and only the report is sent to BM) 2. BM Cloud 3. BM On premise. I am not very sure which of these options are currently supported by DSL. Thank you again, this is awesome and I am going to be extensively using DSL in upcoming months to completely implement CI/CD

**Reply by rabelenda** [reply]
Hello, 

I think I need more info about your listed scenarios (eg: how does each listed option look nowadays for you, what is the general architecture or services involved) to help you, if you need any help with that.

You can run JMeter DSL with JMeter cluster implemented by yourself (using existing JMeter distributed execution support) or with one of the different services the DSL already provides built in integration (currently: BlazeMeter, OctoPerf, Azure Load Testing). [Here](https://abstracta.github.io/jmeter-java-dsl/guide/#run-test-at-scale) are more details.

I am glad you find the DSL useful and exciting and are willing to incorporate it. 

Thank you for letting us know.

**Comment by VaishaliGulve** [comment]
When it comes to testing in BM cloud or on prem, we need to be able to set specific engines since we have dedicated IPs , any options for that? Also, specifying engine names is required to test on premise. 2. We need to be able to specify csv files to be uploaded to BM

**Reply by rabelenda** [reply, marked as answer]
This is great! 

In fact is something we expected users to ask after they start playing further with existing BM engine.

Nowadays the BM engine does not provide options for that but we can easily add them. Both for specifying engines and to upload csv files.

Now that you have requested this, we will prioritize it higher in our backlog. 

Stay tuned for upcoming releases :)

**Reply by rabelenda** [reply]
Do you think that something like this will be enough for the need you describe in this thread?:

```java
testPlan(
    threadGroup(1, 1,
        httpSampler(SAMPLE_LABEL, "https://localhost")
    )
).runIn(new BlazeMeterEngine(System.getenv("BZ_TOKEN"))
    .location("my-location", 2) // location name, location weight
    .location(BlazeMeterEngine.Locations.GCP_US_EAST_1 , 1))
```

**Reply by rabelenda** [reply]
About asset files (CVS), we can consider automatic upload (no need to specify nothing in engine, it will automatically upload files used in csvDataSource) and optional explicit asset upload through new `.asset(file)` method

**Reply by VaishaliGulve** [reply]
that works, let me take a look at csvDataSource 

testPlan(
    threadGroup(1, 1,
        httpSampler(SAMPLE_LABEL, "https://localhost/")
    )
).runIn(new BlazeMeterEngine("projectName", "testName", System.getenv("BZ_TOKEN"))
    .location("my-location", 2) // location name, location weight
    .location(BlazeMeterLocations.GCP_US_EAST_1 , 1)
    .report(“reportName”)
    .files(“list of fileNames”, true/false) // true - splitcsv enable, false - splitcsv disable

**Reply by VaishaliGulve** [reply]
On the jmeter remote testing:

public void testPerformance() throws Exception {
    TestPlanStats stats = testPlan(
        threadGroup(200, Duration.ofMinutes(10),
            httpSampler("http://my.service")
        )
    ).runIn(new DistributedJmeterEngine("BMtestName","host1", "host2")) // test name can be project name, test name
     .report("BMreportName") // the test should be created in specified  BM project
    assertThat(stats.overall().sampleTimePercentile99()).isLessThan(Duration.ofSeconds(5));
  }

}

**Reply by VaishaliGulve** [reply]
Below is an example of one of my taurus.yml config file

modules:
 jmeter:
     properties:
        threads: 135  
        duration: 3600 
        rampup: 300
 blazemeter:
      public-report: true  # set to true to create a public link to the report
      test: XYZ
      project: ABC
execution:
- scenario: PQR
  distributed:
      - IP1:PORT
      - IP2:PORT
scenarios:
  PQR:
   script: ABCDEF.jmx
   modifications:
     disable: # Names of the tree elements to disable
       - Test Element Label1
       - Label2
   
reporting:
- module: final-stats
  dump-xml: stats.xml

**Reply by rabelenda** [reply]
Isn't this taurus scenario with distributed option the same as using [this](https://abstracta.github.io/jmeter-java-dsl/guide/#jmeter-remote-testing). If not, do you know what are the main differences?

**Reply by VaishaliGulve** [reply]
Yes this is the same, with few very imp differences:
1. Ability to send the consolidated report to BM
2. Ability to specify project, test and report name
3. Ability to enable/disable specific tree elements from the yml 

I am not sure if this is already provided by Jmeter DSL and the documentation might not be complete. In that case, pls let me know how I can get full list of features so I can test them.

**Reply by rabelenda** [reply]
I think 1. and 2. should be covered by the new proposed `blazeMeterReportUploader()` element, allowing you to specify token, project, test & report name.

What elements do you usually disable/enable and why? Knowing this info would allow to provide a proper solution for the use case.

**Reply by rabelenda** [reply]
Additionally, just to clarify: do you need to be able to specify locations as the proposal I initially sent (with `location` method), or just using `DistributedJmeterEngine` with host names covers your needs?

**Reply by VaishaliGulve** [reply]
We need both location method as well as DistributedJmeterEngine , we use all 3 test execution methods currently, cloud, on-premise and distributed

**Reply by VaishaliGulve** [reply]
Q: What elements do you usually disable/enable and why? Knowing this info would allow to provide a proper solution for the use case.
Ans: We usually disable/enable samplers or entire thread group ( cases where multiple thread groups are part of single script and might require to enable/disable some of them). Doing this helps us to use a single test script for multiple test scenarios without having to change the original script, or without having to maintain multiple copies of the script
Example: We have campaign offers only at certain part of the year and need to enable the thread group related to campaigns only when they are active, rest of the time they are kept disabled and the rest of the functionality is sufficient to be tested

**Reply by VaishaliGulve** [reply]
Clarification: "I think 1. and 2. should be covered by the new proposed blazeMeterReportUploader() element, allowing you to specify token, project, test & report name."
- Will this option be provided for "DistributedJmeterEngine" as well?

**Reply by rabelenda** [reply]
Considering that JMeter DSL is actually Java code, you can use ifs to include part of test plan or not in certain execution.

For example:

```java
@Test
  public void performanceTest() throws Exception {
    List<TestPlanChild> threadGroups = new ArrayList<>();
    threadGroups.add(
        threadGroup("allWaysOnThreadGroup", ...)
    );
    if ("true".equals(System.getenv("ACTIVE_CAMPAIGNS"))) { // change this with whatever resolution you define for enabling/disabling the element. You might even use some query to DB, or scheduled time check.
      threadGroups.add(
          threadGroup("campaignThreadGroup", ...)
      );
    }
    testPlan(threadGroups.toArray(new TestPlanChild[0]))
        .run();
  }
```

You can even better modularize your test using java methods (eg: creating a `buildAlwaysActiveThreadGroup` & `buildCampaignThreadGroup`), or even classes to be used in different test classes, or even jars included in different java projects.

I think this kind of approach gives you a lot of flexibility, and makes test plan parts more reusable and easier to maintain than having to explicitly change the list of enabled/disabled elements each time.

What do you think?

**Reply by rabelenda** [reply]
Thank you again for taking the time to give all this feedback, and help shaping a new version of JMeter DSL :).

**Reply by VaishaliGulve** [reply]
Yes, completely agree with this :)

**Reply by VaishaliGulve** [reply]
> Considering that JMeter DSL is actually Java code, you can use ifs to include part of test plan or not in certain execution.
> 
> For example:
> 
> ```java
> @Test
>   public void performanceTest() throws Exception {
>     List<TestPlanChild> threadGroups = new ArrayList<>();
>     threadGroups.add(
>         threadGroup("allWaysOnThreadGroup", ...)
>     );
>     if ("true".equals(System.getenv("ACTIVE_CAMPAIGNS"))) { // change this with whatever resolution you define for enabling/disabling the element. You might even use some query to DB, or scheduled time check.
>       threadGroups.add(
>           threadGroup("campaignThreadGroup", ...)
>       );
>     }
>     testPlan(threadGroups.toArray(new TestPlanChild[0]))
>         .run();
>   }
> ```
> 
> You can even better modularize your test using java methods (eg: creating a `buildAlwaysActiveThreadGroup` & `buildCampaignThreadGroup`), or even classes to be used in different test classes, or even jars included in different java projects.
> 
> I think this kind of approach gives you a lot of flexibility, and makes test plan parts more reusable and easier to maintain than having to explicitly change the list of enabled/disabled elements each time.
> 
> What do you think?

Yes! I didnt think of this! This is very powerful! 👍

**Reply by rabelenda** [reply]
> Will this option be provided for "DistributedJmeterEngine" as well?

Yes, we need to review the details and see how we can managed that. But, doing a simple integration, like only sending statistics to Blazemeter once the test ends, should be quite similar to existing `htmlReporter` implementation. If you need realtime reporting (don't remember right now if taurus reporting supports this, would need to review it), that would require more effort, but is something we could analyze.

**Reply by VaishaliGulve** [reply]
> > Will this option be provided for "DistributedJmeterEngine" as well?
> 
> Yes, we need to review the details and see how we can managed that. But, doing a simple integration, like only sending statistics to Blazemeter once the test ends, should be quite similar to existing `htmlReporter` implementation. If you need realtime reporting (don't remember right now if taurus reporting supports this, would need to review it), that would require more effort, but is something we could analyze.

Yes the taurus supports realtime reporting with -report option currently, we use that for our every single test

**Comment by rabelenda** [comment]
Hello, we have just released a [new version of jmeter DSL](https://github.com/abstracta/jmeter-java-dsl/releases/tag/v1.11) which automatically uploads resources to blazemeter and allows to select location. Let me know what do you think.

Regards

**Comment by VaishaliGulve** [comment]
That’s nice , let me test it and get back to you

**Reply by rabelenda** [reply]
Hello, any news? :)

**Reply by VaishaliGulve** [reply]
Hi @rabelenda , Sorry for the late response. I am unable to find the right version of dependency for jmeter-java-dsl-blazemeter, to successfully build my test. What is the version i need to use? I tried the user guide, seems like its down, unable to access it this time.
Thanks

**Reply by VaishaliGulve** [reply]
Here is the error msg: [ERROR] Failed to execute goal on project jmeter-java-dsl-sample: Could not resolve dependencies for project us.abstracta.jmeter:jmeter-java-dsl-sample:jar:0.1-SNAPSHOT: Failed to collect dependencies at us.abstracta.jmeter:jmeter-java-dsl-blazemeter:jar:1.13: Failed to read artifact descriptor for us.abstracta.jmeter:jmeter-java-dsl-blazemeter:jar:1.13: Could not transfer artifact us.abstracta.jmeter:jmeter-java-dsl-blazemeter:pom:1.13 from/to artifactory  authentication failed for us/abstracta/jmeter/jmeter-java-dsl-blazemeter/1.13/jmeter-java-dsl-blazemeter-1.13.pom, status: 401 Unauthorized -> [Help 1]

**Reply by rabelenda** [reply]
Is the guide still down for you? I can access it https://abstracta.github.io/jmeter-java-dsl/guide/#blazemeter

**Reply by rabelenda** [reply]
The version 1.11 includes the new feature, but you can use 1.12 or 1.13 as well.

**Reply by rabelenda** [reply]
The error you are getting seems to be cause by some [Artifactory](https://jfrog.com/artifactory/) instance that you have configured in your maven settings or some proxy in your company redirects to. It looks like authentication is failing with such artifactory instance. Does that make sense?
